src/main/java/com/minecolonies/coremod/commands/AbstractCommandParser.java

```diff
@@ -0,0 +1,675 @@
+            {
+                final List<String> tabCompletions = childs.keySet().stream().filter(k -> k.startsWith(lowerCaseArg0)).collect(Collectors.toList());
+                if (childs.containsKey(lowerCaseArg0))
+                {
+                    final TreeNode<IMenu> childTreeNode = childs.get(lowerCaseArg0);
+                    final IMenu childMenu = childTreeNode.getData();
+                    final ActionMenuState actionMenuState;
+                    if (!childMenu.getMenuType().isNavigationMenu())
+                    {
+                        final ActionMenu actionMenu = (ActionMenu) childMenu;
+                        actionMenuState = new ActionMenuState(actionMenu);
+                    }
+                    else
+                    {
+                        actionMenuState = null;
+                    }
+                    return new ParsingResult(tabCompletions, childTreeNode, Collections.emptyList(), actionMenuState, (String) null);
+                }
+                else
+                {
+                    return new ParsingResult(tabCompletions, treeNode, Collections.emptyList(), (ActionMenuState) null, lowerCaseArg0);
+                }
+            }
+            final TreeNode<IMenu> child = childs.get(lowerCaseArg0);
+            if (null == child)
+            {
+                final List<String> tabCompletions = Collections.emptyList();
+                return new ParsingResult(tabCompletions, treeNode, Collections.emptyList(), (ActionMenuState) null, lowerCaseArg0);
+            }
+            final String[] newArgs = new String[args.length - 1];
+            System.arraycopy(args, 1, newArgs, 0, newArgs.length);
+            return getTabCompletionsAndParsingHolders(child, server, sender, newArgs, pos, moduleContext);
+        }
+        else
+        {
+            final List<ActionMenuHolder> parsedHolders = new ArrayList<>();
+            final Map<String, ActionMenuHolder> possibleActionCommands = getActionCommands(treeNode, parsedHolders);
+            @NotNull final List<ActionArgument> parsedActionArgumentList = new ArrayList<>();
+            final ActionMenu actionMenu = (ActionMenu) treeNode.getData();
+            @NotNull final ActionMenuState actionMenuState = new ActionMenuState(actionMenu);
+            return getTabCompletionsAndParsingHoldersForActionMenuTreeNode(treeNode, actionMenuState, parsedHolders, parsedActionArgumentList, possibleActionCommands, server,
+                    sender, args, pos, moduleContext);
+        }
+    }
+
+    @NotNull
+    private ParsingResult getTabCompletionsAndParsingHoldersForActionMenuTreeNode(
+                                           @NotNull final TreeNode<IMenu> actionMenuTreeNode,
+           // TODO: can parsedHolders, parsedActionArgumentList, and possibleActionCommands be merged into actionMenuState?
+                                           @NotNull final ActionMenuState actionMenuState,
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)